### DISCUSSION TITLE
"Enabled No Discharge" mode error by modbus power control - please help

### DISCUSSION THREAD

**Discussion by autacek** [wills106/homeassistant-solax-modbus#672] [General]
The option works, but the power from PV to GRID is stoped too and production limited. How is the right solution? Thank you. 

P.S. I need only no discharge battery, all other direction without limits (or as by Disabled mode).

**Comment by infradom** [comment]
There might indeed be an error in the code of plugin_solax.py . I use no_discharge, but seldom in situations with excess PV, so this may not have been tested sufficiently.

As this stuff is fairly confusing, we need to be carefull making modifications without second thought.
So the section
`    elif power_control == "Enabled No Discharge": # alternative computation for Power Control
        if pv <= houseload_nett: ap_target = 0 - pv + (houseload_brut - houseload_nett) # 0 - pv + (houseload_brut - houseload_nett)
        else:                    ap_target = 0 - houseload_nett`

could probably be replaced by:
`    elif power_control == "Enabled No Discharge": # alternative computation for Power Control
        if pv <= houseload_nett: ap_target = 0 - pv + (houseload_brut - houseload_nett) # 0 - pv + (houseload_brut - houseload_nett)
        else:                    ap_target = 0 - pv - houseload_nett`

I will give it some thought and maybe test on my system.

**Reply by infradom** [reply]
I implemented this already in my development system/fork now. I need a sunny day to be sure it works as expected.
If it works as expected, I will issue a PR to the @wills106 main line.

**Reply by autacek** [reply]
I have a similar problem with powercontrol (enabled battery control) as well. Practically, after charging the battery, production from PV will be limited. Then the situation differs according to the Remotecontrol Active Power value from -10000 (forced supply to the grid) to +10000 (forced battery charging), if it is 0, then the battery is not charged and the PV production goes to the grid. Logically, there is no problem at night, but during daylight it substantially limits the production to only the consumption of the house. Experimenting with Remotecontrol Import Limit also does not lead to results.

**Comment by infradom** [comment]
Just an additional note: my change did not work as expected, so I have undone that change. 
Sorry for the delay in finding a real solution. I am very busy these days with a renovation of my son's house.
Hope to find some time soon ...
I myself run the enabled_no_discharge mode only in certain circumstances, so an automation stops this mode if the battery is (nearly)  full.
